src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.MakeFieldReadonly
+{
+    internal abstract class AbstractMakeFieldReadonlyDiagnosticAnalyzer<TIdentifierNameSyntax, TConstructorDeclarationSyntax>
+        : AbstractCodeStyleDiagnosticAnalyzer
+        where TIdentifierNameSyntax : SyntaxNode
+        where TConstructorDeclarationSyntax : SyntaxNode
+    {
+        protected AbstractMakeFieldReadonlyDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)))
+        {
+        }
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
```

Comment:
Analyzer using `ISyntaxFactsService` seems extremely fishy, and will likely cause issues when we try to remove the `IBuildInAnalyzer` hack that exposes `Workspace` to the IDE analyzers.

Reply:
➡️ Resolved by filing #25652

Reply:
Tons of analyzers use ISyntaxFactsService.    I'm not sure why this is a problem.

>and will likely cause issues when we try to remove the IBuildInAnalyzer hack that exposes Workspace to the IDE analyzers.

We don't need the workspace to get at these services.  They're effectively singletons.

Reply:
> Tons of analyzers use ISyntaxFactsService

Unless the analyzer is doing extremely workspace specific analysis, it should not be using any workspace services, otherwise we can never port this analyzer to run in command line builds.

Reply:
> Unless the analyzer is doing extremely workspace specific analysis, it should not be using any workspace services

ISyntaxFactsService is not an IWorkspaceService it's an ILanguageService.  It exists so we can abstract away differences between the languages.

We use lots of ILanguageServices in analyzers because, unsurprisingly, we have written lots of glue code that works across languages, and we want to be able to share them in analyzers.

Reply:
Note that we should have no problem using ISyntaxFactsService from the command line.  It's literally an aggregation of static helpers, wrapped into an interface so it can be called uniformly across VB and C#.  It's the moral equivalent of SyntaxFacts+SyntaxFactory, just usable from both languages.  It should be 100% fine to use now, and even if we move these to be command line analyzers.

Reply:
Would it be worthwhile to ship these publically as analyzer utilities? I recall we have talked about this in the past when we started working on roslyn-analyzers, and we took the easier route of just copy-paste useful helpers/extension methods from workspaces layer to the analyzers repo.

---
In reply to: [176218519](https://github.com/dotnet/roslyn/pull/19067#discussion_r176218519) [](ancestors = 176218519)

Reply:
> Would it be worthwhile to ship these publically as analyzer utilities? 

Sure.  As long as that doesn't negatively impede teh IDE's ability to make changes easily, and as long as it doesn't lock the IDE into having to support a public API.  These APIs underpin tons of our features, and we need to be able to efficiently and effectively make updates here.  If that isn't impacted, then i see no problem making this available to others.